physicsnemo/metrics/diffusion/sfm_loss.py

```diff
@@ -0,0 +1,251 @@
+        sigma_min: Union[List[float], float], optional
+            Minimum value of noise sigma, default 2e-3
+            Protects against values near zero that result in loss explosion.
+        sigma_data: float, optional
+            EDM weighting, default 0.5
+        """
+        self.encoder_loss_type = encoder_loss_type
+        self.encoder_loss_weight = encoder_loss_weight
+        self.sigma_min = sigma_min
+        self.sigma_data = sigma_data
+
+        if encoder_loss_type not in ["l1", "l2", None]:
+            raise ValueError(
+                f"encoder_loss_type should be one of ['l1', 'l2', None] not {encoder_loss_type}"
+            )
+
+    def __call__(
+        self,
+        denoiser_net: torch.nn.Module,
+        encoder_net: torch.nn.Module,
+        img_clean: torch.Tensor,
+        img_lr: torch.Tensor,
+    ):
+        """
+        Calculate the loss for corresponding to stochastic flow matching
+
+        Parameters
+        ----------
+            denoiser_net: torch.Tensor
+                The denoiser network making the predictions
+            encoder_net: torch.Tensor
+                The encoder network making the predictions
+            img_clean: torch.Tensor
+                Input images (high resolution) to the neural network.
+            img_lr: torch.Tensor
+                Input images (low resolution) to the neural network.
+        Returns
+        -------
+            torch.Tensor
+            A tensor representing the combined loss calculated based on the flow matching
+            encoder and denoiser networks
+        """
+        # uniformly samples from 0 to 1 in torch
+        if isinstance(denoiser_net, torch.nn.parallel.DistributedDataParallel):
+            sigma_max_per_channel = denoiser_net.module.get_sigma_max().to(
+                device=img_clean.device
+            )
+        else:
+            sigma_max_per_channel = denoiser_net.get_sigma_max().to(
+                device=img_clean.device
+            )
+
+        # clamp to min value
+        if not isinstance(self.sigma_min, float) and len(self.sigma_min) > 1:
+            sigma_max_per_channel = torch.maximum(
+                sigma_max_per_channel,
+                torch.tensor(self.sigma_min, device=img_clean.device),
+            )
+            # Normalize from 0 to 1
+            sigma_max = 1.0
+        else:
+            # just use the first value, ignore the rest
+            sigma_max = torch.maximum(
+                sigma_max_per_channel,
+                torch.tensor(self.sigma_min, device=img_clean.device),
+            )[0]
+
+        rnd_uniform = torch.rand([img_clean.shape[0], 1, 1, 1], device=img_clean.device)
+
+        sampled_sigma = rnd_uniform * sigma_max
+        weight = (sampled_sigma**2 + self.sigma_data**2) / (
+            sampled_sigma * self.sigma_data
+        ) ** 2
+
+        # augment for conditional generaiton
+        x_tot = torch.cat((img_clean, img_lr), dim=1)
```

Comment:
This concat has no real point, right? `x_tot` is immediately split out back into `x_1` and `x_low` and then never accessed again. Also, what does the note about augmenting and conditional generation mean?